vita3k/gui/src/ime.cpp

```diff
@@ -42,129 +74,391 @@ enum CapsLevel {
-    ImGui::PushStyleVar(ImGuiStyleVar_FrameRounding, 10.f);
-    ImGui::Columns(9, nullptr, false);
-    ImGui::PushStyleColor(ImGuiCol_Button, GUI_COLOR_TEXT);
-    ImGui::PushStyleColor(ImGuiCol_Text, GUI_COLOR_TEXT_BLACK);
-    for (const auto &key : host.ime.caps_level == NO ? key : shift_key) {
-        if (ImGui::Button(key.c_str(), BUTTON_SIZE) && (host.ime.str.size() < host.ime.param.maxTextLength))
-            update_str(host.ime, key);
-        ImGui::NextColumn();
+static void update_key(Ime &ime, const std::u16string &key) {
+    update_str(ime, key);
+    ime.edit_text.editLengthChange = ime.edit_text.preeditLength += (int32_t)key.length();
+}
+
+static void reset_preedit(Ime &ime) {
+    ime.caretIndex = ime.edit_text.caretIndex;
+    ime.edit_text.preeditIndex = ime.edit_text.caretIndex;
+    ime.edit_text.preeditLength = ime.edit_text.editLengthChange = 0;
+    ime.event_id = SCE_IME_EVENT_UPDATE_TEXT;
+}
+
+static void update_ponct(Ime &ime, const std::u16string &key) {
+    update_str(ime, key);
+    reset_preedit(ime);
+}
+
+static std::map<int, float> lang_keyboard_pos;
+static std::map<int, std::vector<std::u16string>> lang_key, shift_lang_key, ponct;
+static std::string space_str;
+
+void init_ime_lang(Ime &ime, const SceImeLanguage &lang) {
+    // Set Ponctuation
+    switch (lang) {
+    case SCE_IME_LANGUAGE_SPANISH:
+        break;
+    case SCE_IME_LANGUAGE_DANISH:
+    case SCE_IME_LANGUAGE_GERMAN:
+    case SCE_IME_LANGUAGE_ENGLISH_GB:
+    case SCE_IME_LANGUAGE_ENGLISH_US:
+    case SCE_IME_LANGUAGE_FRENCH:
+    case SCE_IME_LANGUAGE_ITALIAN:
+    case SCE_IME_LANGUAGE_DUTCH:
+    case SCE_IME_LANGUAGE_NORWEGIAN:
+    case SCE_IME_LANGUAGE_POLISH:
+    case SCE_IME_LANGUAGE_PORTUGUESE_BR:
+    case SCE_IME_LANGUAGE_PORTUGUESE_PT:
+    case SCE_IME_LANGUAGE_FINNISH:
+    case SCE_IME_LANGUAGE_SWEDISH:
+    case SCE_IME_LANGUAGE_TURKISH:
+    default:
+        ponct = { { 1, { u".", u"?" } }, { 2, { u",", u"!" } } };
+        break;
     }
-    ImGui::PopStyleColor(2);
-    ImGui::Columns(1);
-    ImGui::SameLine(0, 15);
-    ImGui::PushStyleColor(ImGuiCol_Button, GUI_IME_BUTTON_BG);
-    if (ImGui::Button("Back", BUTTON_SIZE) && (host.ime.edit_text.caretIndex > 0)) {
-        host.ime.str.erase(host.ime.edit_text.caretIndex - 1, 1);
-        host.ime.edit_text.editIndex = host.ime.edit_text.caretIndex;
-        --host.ime.edit_text.caretIndex;
-        if (host.ime.edit_text.preeditIndex > host.ime.edit_text.caretIndex)
-            host.ime.edit_text.preeditIndex = host.ime.caretIndex = host.ime.edit_text.caretIndex;
-        host.ime.edit_text.editLengthChange = -1;
-        if (host.ime.edit_text.preeditLength > 0)
-            --host.ime.edit_text.preeditLength;
-        host.ime.event_id = SCE_IME_EVENT_UPDATE_TEXT;
-        if (host.ime.str.empty() && (host.ime.caps_level == NO))
-            host.ime.caps_level = YES;
+    // Set Space String
+    switch (lang) {
```

Comment:
Maybe a map is better for this

Reply:
no understund map ? i ussing diferent switch because for some thinks is same other and for some is diferent